dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -861,22 +862,28 @@ else if ( OrganisationUnitSelectionMode.CHILDREN.equals( orgUnitSelectionMode )
 
         if ( user != null )
         {
+            Set<OrganisationUnit> orgUnits = new HashSet<>();
+
             if ( OrganisationUnitSelectionMode.ACCESSIBLE.equals( orgUnitSelectionMode ) )
             {
-                Set<OrganisationUnit> orgUnits = program.isRegistration()
+                orgUnits = program.isRegistration()
                     ? user.getTeiSearchOrganisationUnitsWithFallback()
                     : user.getDataViewOrganisationUnitsWithFallback();
-
-                organisationUnits.addAll( organisationUnitService.getOrganisationUnitsWithChildren( orgUnits.stream()
-                    .map( BaseIdentifiableObject::getUid ).collect( Collectors.toList() ) ) );
             }
             else if ( OrganisationUnitSelectionMode.CAPTURE.equals( orgUnitSelectionMode ) )
             {
-                organisationUnits.addAll(
-                    organisationUnitService.getOrganisationUnitsWithChildren( user.getOrganisationUnits().stream()
-                        .map( BaseIdentifiableObject::getUid ).collect( Collectors.toList() ) ) );
+                orgUnits = user.getOrganisationUnits();
+            }
+            else if ( OrganisationUnitSelectionMode.ALL.equals( orgUnitSelectionMode )
+                && (user.isSuper() || currentUserService
+                    .currentUserIsAuthorized( Authorities.F_TRACKED_ENTITY_INSTANCE_SEARCH_IN_ALL_ORGUNITS.name() )) )
+            {
+                orgUnits = new HashSet<>( organisationUnitService.getRootOrganisationUnits() );
             }
 
+            organisationUnits.addAll( organisationUnitService.getOrganisationUnitsWithChildren( orgUnits.stream()
```

Comment:
Wondering what happens if orgUnits is still empty in this line because ouMode is neither of ACCESSIBLE/CAPTURE/ALL. Assuming that there wont be an "empty" query going to the db and its verified?

Reply:
if orgUnits is empty, empty list would be passed. We check if orgUnits is null or empty while generating the query.... which for me is the best place to check.

Reply:
Another case is when ALL is selected, but user does not have the authority and is not superuser; We would end up with an empty list. I am wondering if we should fallback to ACCESSIBLE, throw an error, or if, infact, returning an empty list is the solution; What do you think @abyot, @Bekkalizer ?

Reply:
@stian-sandvold, if user asks for ALL without having the authority, we are rejecting the request.